drv/ignition-api/src/lib.rs

```diff
@@ -33,105 +36,395 @@ pub const PORT_MAX: usize = 40;
+        self.controller.port_count()
+    }
+
+    /// Return a u64 with each bit indicating whether or not a Target is present
+    /// for this port. See the note above on `PORT_MAX` about the upper bound.
+    #[inline]
+    pub fn presence_summary(&self) -> Result<u64, IgnitionError> {
+        self.controller.presence_summary()
+    }
+
+    /// Return the raw state for the given port. The internals of `PortState`
+    /// are purposefully opague since they are closely tied to the RTL
+    /// implementation and as such are not intended for direct consumption. The
+    /// `port(..)` and `target(..)` functions below are probably what you want
+    /// to use.
+    #[inline]
+    pub fn raw_port_state(&self, port: u8) -> Result<PortState, IgnitionError> {
+        self.controller.port_state(port)
+    }
+
+    /// Return the state for the given port.
+    pub fn port(&self, port: u8) -> Result<Port, IgnitionError> {
+        self.raw_port_state(port).map(Port::from)
+    }
+
+    /// Return the `Target` for a given port if present.
+    pub fn target(&self, port: u8) -> Result<Option<Target>, IgnitionError> {
+        self.raw_port_state(port)
+            .map(|state| Port::from(state).target)
+    }
+
+    /// Send the given system power `Request` to the given port. Once a request
+    /// is sent and accepted by the Target, it enforces a cooldown before
+    /// subsequent requests are accepted and processed. `SystemPowerOff` and
+    /// `SystemPowerOn` requests have a three second cooldown while a
+    /// `SystemPowerReset` has a six seconds cooldown.
+    #[inline]
+    pub fn send_request(
+        &self,
+        port: u8,
+        request: Request,
+    ) -> Result<(), IgnitionError> {
+        self.controller.send_request(port, request)
+    }
+
+    /// Return the `Counters` for the given port. This function has the
+    /// side-effect of clearing the counters.
+    #[inline]
+    pub fn counters(&self, port: u8) -> Result<Counters, IgnitionError> {
+        self.controller.counters(port)
+    }
+
+    /// Return the `LinkEvents` for the given transceiver and port. See
+    /// `TransceiverSelect` for more details.
+    pub fn link_events(
+        &self,
+        port: u8,
+        txr: TransceiverSelect,
+    ) -> Result<LinkEvents, IgnitionError> {
+        self.controller.link_events(port, txr).map(LinkEvents::from)
+    }
+
+    /// Clear the link events for the given transceiver and port. See
+    /// `TransceiverSelect` for more details.
+    pub fn clear_link_events(
+        &self,
+        port: u8,
+        txr: TransceiverSelect,
+    ) -> Result<(), IgnitionError> {
+        self.controller.clear_link_events(port, txr)
+    }
+
+    /// Return the state of all ports in a single operation. Be aware that this
+    /// reply is fairly large and may require enlarging the stack of the caller.
+    pub fn all_ports(
+        &self,
+    ) -> Result<[Port; PORT_MAX as usize], IgnitionError> {
```

Comment:
Thinking out loud - this (and `all_link_events()` below) always return a `PORT_MAX`-length array, which the caller must then shorten by knowing the actual port count. Would it make sense to return an iterator instead? If so, I think this could even return a concrete iterator type without having to write one by hand. Untested, but roughly:

```rust
pub fn all_ports(&self) -> Result<core::iter::Take<core::array::IntoIter<Port, PORT_MAX>>, IgnitionError> {
    let port_count = usize::from(self.port_count()?);
    let all_ports = self.controller.all_port_state()?.map(Port::from);
    Ok(all_ports.into_iter().take(port_count))
}
```

Reply:
Hmm.. I like this a lot and should have thought of this since it avoids copying from one array to another as data goes from `PortState` to `Port`. It should reduce stack pressure in the caller and I may even double down on this in a follow up and allow requesting these in two or four batches instead of one large array. This would fit better with the model I need for Hiffy.

Reply:
The version I wrote wouldn't reduce stack pressure on the caller, since the `all_ports` array gets moved into the iterator that's returned. But if you do some rework to use iterators or batched calls all the way down, it certainly could.